klippy/extras/load_cell_probe.py

```diff
@@ -0,0 +1,699 @@
+        speed = self._param_helper.get_probe_params(gcmd)['probe_speed']
+        phoming = self._printer.lookup_object('homing')
+        # start collector after tare samples are consumed
+        collector = self._start_collector()
+        # do homing move
+        return phoming.probing_move(self, pos, speed), collector
+
+    # Wait for the MCU to trigger with no movement
+    def probing_test(self, gcmd, timeout):
+        self._pause_and_tare(gcmd)
+        toolhead = self._printer.lookup_object('toolhead')
+        print_time = toolhead.get_last_move_time()
+        self._home_start(print_time)
+        return self.home_wait(print_time + timeout)
+
+    def get_status(self, eventtime):
+        return {
+            'tare_counts': self._tare_counts,
+            'last_trigger_time': self._last_trigger_time,
+        }
+
+
+# Perform a single complete tap
+class TappingMove:
+    def __init__(self, config, load_cell_probing_move,
+                 config_helper):
+        self._printer = config.get_printer()
+        self._load_cell_probing_move = load_cell_probing_move
+        self._config_helper = config_helper
+        # track results of the last tap
+        self._last_result = None
+        self._is_last_result_valid = False
+        # webhooks support
+        self._clients = load_cell.ApiClientHelper(config.get_printer())
+        name = config.get_name()
+        header = {"header": ["probe_tap_event"]}
+        self._clients.add_mux_endpoint("load_cell_probe/dump_taps",
+                                      "load_cell_probe", name, header)
+
+    # perform a probing move and a pullback move
+    def run_tap(self, gcmd):
+        # do the descending move
+        epos, collector = self._load_cell_probing_move.probing_move(gcmd)
+        # collect samples from the tap
+        toolhead = self._printer.lookup_object('toolhead')
+        toolhead.flush_step_generation()
+        move_end = toolhead.get_last_move_time()
+        results = collector.collect_until(move_end)
+        samples = check_sensor_errors(results, self._printer)
+        # Analyze the tap data
+        ppa = TapAnalysis(samples)
+        # broadcast tap event data:
+        self._clients.send({'tap': ppa.to_dict()})
+        self._is_last_result_valid = True
+        self._last_result = epos[2]
+        return epos, self._is_last_result_valid
+
+    def get_status(self, eventtime):
+        return {
+            'last_z_result': self._last_result,
+            'is_last_tap_valid': self._is_last_result_valid
+        }
+
+
+# Probe `activate_gcode` and `deactivate_gcode` support
+class ProbeActivationHelper:
+    def __init__(self, config):
+        self._printer = config.get_printer()
+        gcode_macro = self._printer.load_object(config, 'gcode_macro')
+        self._activate_gcode = gcode_macro.load_template(
+            config, 'activate_gcode', '')
+        self._deactivate_gcode = gcode_macro.load_template(
+            config, 'deactivate_gcode', '')
```

Comment:
Please do not introduce `activate_gcode` / `deactivate_gcode` for the initial load cell implementation.  I fear they will conflict with upcoming homing/probing rework.  I'm leery of introducing new config settings just to deprecate them a few weeks later.

I understand the desire to use these callbacks to introduce temperature sanity checks.  Until we can get a better feel for the ultimate homing/probing API I would request that users perform these temperature tests as part of the macros that call the probing commands (instead of as macros that get called *from* the probing commands).

Reply:
I'll move this into the next PR.

This sounds discouraging. Users do things from the front end terminal, not just their start gcode. Things like `PROBE`, `PROBE_ACCURACY`, `BED_MESH`, `LOAD_CELL_TAP`, `QUAD_GANTRY_LEVEL` and/or `Z_TILT_ADJUST` (did I miss some?). Making all of that safe one by one feels like an unnecessary and complicated ask.

FWIW, `activate_gcode` / `deactivate_gcode` already exist on probes.